### PR TITLE
Extragradient and tests

### PR DESCRIPTION
**Pull request by manuel-delverme**
(This pull request depends on the HS suite)
adds:

1. Extragradient optimizer and tests (18/20 passing)
2. gradient descent/ascent, untested
3. updates in the pip package
4. a bug fix (?) https://github.com/gehring/fax/compare/master...manuel-delverme:extragradient_test?expand=1#diff-d6645bfd4bdc4d6e51625269e2018617R106

3 and 4 should be part of another PR, let me know if i should split this

### CODE REVIEW

**Comment by gehring** [comment]
Can you open an issue with a brief explanation of what the bug is. It's not obvious to me from your link.

Could you elaborate on what is preventing us from passing 2/20 tests?

File: fax/competitive/cga.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import collections
 from functools import partial
+from typing import Tuple
```
Comment (pierrelux) [inline fax/competitive/cga.py:3]: Since this is not related to extragradient, perhaps better to slice into separate PR

File: fax/competitive/sgd.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+from typing import Callable
```
Comment (pierrelux) [inline fax/competitive/sgd.py:1]: Perhaps better to slice into separate PR, just for GDA

File: fax/constrained/constrained.py [hunk 3]
```diff
@@ -225,7 +225,8 @@ def lagrange_update(i, grads, opt_state, *args, **kwargs):
             An new packed optimization state with the updated parameters and
             Lagrange multipliers.
         """
-        grads = (grads[0], tree_util.tree_map(lax.neg, grads[1]))
+        params_grad, multipliers_grad = grads
```
Comment (pierrelux) [inline fax/constrained/constrained.py:228]: Just aesthetical?

File: fax/constrained/constrained.py [hunk 4]
```diff
@@ -304,7 +305,8 @@ def _equality_constraints(variables):
 
     @jit
     def update(i, opt_state):
-        grads = grad(lagrangian, (0, 1))(*get_params(opt_state))
+        grad_fn = grad(lagrangian, (0, 1))
```
Comment (pierrelux) [inline fax/constrained/constrained.py:308]: Aesthetical?

File: fax/constrained/constrained_test.py [hunk 5]
```diff
@@ -64,29 +71,29 @@ def step(i, opt_state):
...
+        {'method': slsqp_ecp, 'kwargs': {'max_iter': 1000}}, )
     @hypothesis.settings(max_examples=10, deadline=5000.)
     @hypothesis.given(
         hypothesis.extra.numpy.arrays(
             onp.float, (2,),
             elements=hypothesis.strategies.floats(0.1, 1)),
     )
     def test_ecp(self, method, kwargs, v):
-        opt_solution = (1./np.linalg.norm(v))*v
+        opt_solution = (1. / np.linalg.norm(v)) * v
```
Comment (pierrelux) [inline fax/constrained/constrained_test.py:87]: Just formatting? We don't have strong guidelines for formatting, but should perhaps settle on this to avoid having to much diff to parse

File: fax/loop.py [hunk 6]
```diff
@@ -104,20 +103,19 @@ def fixed_point_iteration(init_x, func, convergence_test, max_iter,
 
     def cond(args):
         i, x_new, x_old = args
+        x_new, x_old = get_params(x_new), get_params(x_old)
```
Comment (pierrelux) [inline fax/loop.py:106]: Was that a bug?

File: fax/loop.py [hunk 7]
```diff
@@ -104,20 +103,19 @@ def fixed_point_iteration(init_x, func, convergence_test, max_iter,
...
 
         if max_iter is not None:
             converged = converged | (max_iter <= i)
         return np.logical_not(converged)
 
     def body(args):
-        i, x_new, _ = args
-        i_new, x_new, x_old = unrolled(i, x_new, func, batched_iter_size,
-                                       return_last_two=True)
+        i, x_new, _x_old = args
```
Comment (pierrelux) [inline fax/loop.py:114]: Same comment as above: would be good not to interfere with the formatting too much.

File: fax/loop.py [hunk 8]
```diff
@@ -151,3 +149,65 @@ def scan_step(args, idx):
...
+
+    solution = fixed_point_iteration(init_x, func, convergence_test, max_iter, batched_iter_size, unroll, get_params)
+
+    jax.lax.while_loop = jax_while_loop
+
+    plot_process(js, xs, ys)
+    return solution
+
+
+def plot_process(js, xs, ys):
```
Comment (pierrelux) [inline fax/loop.py:197]: We shouldn't have plotting code in there

File: fax/loop.py [hunk 9]
```diff
@@ -151,3 +149,65 @@ def scan_step(args, idx):
         iterations=iterations,
         previous_value=prev_sol,
     )
+
+
+def _debug_fixed_point_iteration(init_x, func, convergence_test, max_iter, batched_iter_size=1,
```
Comment (pierrelux) [inline fax/loop.py:154]: Unclear if this should go in this file. @gehring would know better. Also @NeilGirdhar proposed a scan-based idea which may encompass this. Overall, I suggest putting all changes to loop.py in a separate PR since unrelated to extragradient

File: fax/loop_test.py [hunk 10]
```diff
@@ -80,13 +79,13 @@ def step(i, x_old):
...
 
         sol = loop.fixed_point_iteration(
-                init_x=init_x,
-                func=step,
-                convergence_test=convergence_test,
-                max_iter=max_steps,
-                batched_iter_size=1,
-                unroll=unroll,
-            )
+            init_x=init_x,
```
Comment (pierrelux) [inline fax/loop_test.py:82]: Formatting comment

File: fax/test_util.py [hunk 11]
```diff
@@ -188,7 +195,216 @@ def func(u):
...
+            rest, num = a.split("[")
+            b = f"{rest}[{int(num) - 1}"
+            seq.append(b)
+    cost_function = "]".join(seq)
+    scope = ", ".join(k for k in closure.keys() if not k.startswith("__"))
+    cost_function_ = f"{variable} = lambda {scope}: {cost_function}"
+    return cost_function_
+
+
+def apm_to_python(text: Text) -> Union[Text, None]:
```
Comment (pierrelux) [inline fax/test_util.py:268]: Probably ok for now, but perhaps at some point would be good to clean up test_util.py and separate things out into different files

File: setup.py [hunk 12]
```diff
@@ -1,18 +1,39 @@
-from setuptools import setup, find_namespace_packages
+import setuptools
```
Comment (pierrelux) [inline setup.py:1]: This should go into a separate PR

File: fax/test_util.py [hunk 13]
```diff
@@ -188,7 +195,216 @@ def func(u):
...
+    """Convert APM format to a python code file.
```
Comment (pierrelux) [inline fax/test_util.py:269]: This is a duplicate of #14

**Comment by pierrelux** [comment]
Thanks for your work. General comment: this PR should really be split into smaller ones. Also, there seems to be a duplicate of code with #14. I also realized that #14 changes the CGA-related code to add type annotation and other auto-formatting changes: I would keep that separate as well and we should generally try to configure our IDEs so that we don't re-format everything on save.

**Comment by manuel-delverme** [comment]
> 2/20 tests
Here I'm using a set of hyperparameters for all the examples,
I think i would have to create per-test hyperparameters to reach 20/20, it did not seem important

File: fax/loop.py (hunk 6 above)
Comment (manuel-delverme) [reply fax/loop.py:106]: see #22

File: fax/loop.py (hunk 9 above)
Comment (manuel-delverme) [reply fax/loop.py:154]: reverting these changes, I should have removed the viz code.

File: fax/constrained/constrained.py (hunk 3 above)
Comment (manuel-delverme) [reply fax/constrained/constrained.py:228]: It implicitly enforces the `grads` structure as the docstring requires
in the indexing version `grads = [1, 2, 3]` would be acceptable, now it fails.
I'm removing it nonetheless because it's not an extragrad change

File: fax/constrained/constrained.py (hunk 4 above)
Comment (manuel-delverme) [reply fax/constrained/constrained.py:308]: it allows for exception messages to be more detailed, both the `grad` and grad evaluation are complex functions.
reverting for the same reasoning as above

File: fax/constrained/constrained_test.py (hunk 5 above)
Comment (manuel-delverme) [reply fax/constrained/constrained_test.py:87]: only formatting, the reason of the random -> jax.random is that especially during development using jax rng is a big overhead so i have used python's random

File: fax/loop.py (hunk 7 above)
Comment (manuel-delverme) [reply fax/loop.py:114]: reverted

File: fax/loop.py (hunk 9 above)
Comment (NeilGirdhar) [reply fax/loop.py]: In case you're curious, here's [my version](https://github.com/NeilGirdhar/fax_suggestions/blob/master/cmm/fixed_point/iterated_function.py#L81).   Using scan is much faster since it compiles the loop.   The tap function is a new JAX feature that lets us have a progress bar that operates during the scan.  The extract function lets us get information out of the scan.  Also, I have the parameters as a regular object rather than a closure, which allows the parameters to contain tracers and be compiled.  Without this design, you cannot nest this call in another compiled function.

Comment (gehring) [reply fax/loop.py]: @NeilGirdhar As far as I understand, `jax.lax.while_loop` should be compiled in the same situations as `jax.lax.scan`. Scan can possibly have better memory allocation but I'm not sure how well XLA leverages that. From the `jax.lax.scan` docs, "also unlike that Python version, scan is a JAX primitive and is lowered to a single XLA While HLO."

In what setting have you encountered significantly different performance?

Comment (NeilGirdhar) [reply fax/loop.py]: @gehring I guess you're responding to my statement about "scan being much faster".  I was comparing using scan to Manuel's version that uses a plain Python for loop to sample a trajectory.  I guess that wasn't clear.

Comment (gehring) [reply fax/loop.py]: Ah, my bad! Then yes, I agree.

Though, when jitted, the speed up will mostly be in compile time. In fact, I believe there are some cases where a statically compiled python loop will be faster than scan due to how stuff gets dispatched to the gpu (similar to inefficiencies with while loops having to check with the cpu at every iteration). I recall reading some open issue about it so presumably the XLA team will be improving this at some point.

Small aside, this overhead with GPU + loops is why I added the batching of iterations to the implementation of the fixed point iteration loop.

Comment (NeilGirdhar) [reply fax/loop.py]: Interesting!

**Comment by gehring** [comment]
@pierrelux Have all your requested changes been addressed?
